components/homme/cmake/HommeMacros.cmake

```diff
@@ -172,6 +175,94 @@ macro(createTestExec execName execType macroNP macroNC
 
 endmacro(createTestExec)
 
+# Create a library instead of an executable, so Homme can be used by
+# another cmake project as a dycore library
+macro(createExecLib libName execType libSrcs inclDirs macroNP
+                    macroPLEV macroWITH_ENERGY macroQSIZE_D)
+
+  # Set the include directories
+  SET(modulesDir "${CMAKE_CURRENT_BINARY_DIR}/${libName}_modules")
+
+  MESSAGE(STATUS "Building ${libName} library derived from ${execType} with:")
+  MESSAGE(STATUS "  NP = ${macroNP}")
+  MESSAGE(STATUS "  PLEV = ${macroPLEV}")
+  MESSAGE(STATUS "  QSIZE_D = ${macroQSIZE_D}")
+  MESSAGE(STATUS "  ENERGY = ${macroWITH_ENERGY}")
+
+  # Set the variable to the macro variables
+  SET(NUM_POINTS ${macroNP})
+  SET(NUM_PLEV ${macroPLEV})
+
+  IF (${macroWITH_ENERGY})
+    SET(ENERGY_DIAGNOSTICS TRUE)
+  ELSE()
+    SET(ENERGY_DIAGNOSTICS)
+  ENDIF ()
+
+  IF (${macroQSIZE_D})
+    SET(QSIZE_D ${macroQSIZE_D})
+  ELSE()
+    SET(QSIZE_D)
+  ENDIF ()
+
+  # This is needed to compile the test executables with the correct options
+  SET(THIS_CONFIG_IN ${HOMME_SOURCE_DIR}/src/${execType}/config.h.cmake.in)
+  SET(THIS_CONFIG_HC ${CMAKE_CURRENT_BINARY_DIR}/config.h.c)
+  SET(THIS_CONFIG_H ${CMAKE_CURRENT_BINARY_DIR}/config.h)
+
+  # First configure the file (which formats the file as C)
+  HommeConfigFile (${THIS_CONFIG_IN} ${THIS_CONFIG_HC} ${THIS_CONFIG_H} )
+
+  ADD_DEFINITIONS(-DHAVE_CONFIG_H)
+
+  ADD_LIBRARY(${libName} ${libSrcs})
+  TARGET_INCLUDE_DIRECTORIES (${libName} PUBLIC ${inclDirs} ${modulesDir} ${CMAKE_CURRENT_BINARY_DIR})
+  SET_TARGET_PROPERTIES(${libName} PROPERTIES Fortran_MODULE_DIRECTORY ${modulesDir})
+  SET_TARGET_PROPERTIES(${libName} PROPERTIES LINKER_LANGUAGE Fortran)
+  IF(BUILD_HOMME_WITHOUT_PIOLIBRARY)
+    TARGET_COMPILE_DEFINITIONS(${libName} PUBLIC HOMME_WITHOUT_PIOLIBRARY)
+  ENDIF()
+
+
+  IF (CXXLIB_SUPPORTED_CACHE)
+    MESSAGE(STATUS "   Linking Fortran with -cxxlib")
+    TARGET_LINK_LIBRARIES(${libName} -cxxlib)
+  ENDIF ()
+
+  STRING(TOUPPER "${PERFORMANCE_PROFILE}" PERF_PROF_UPPER)
+  IF ("${PERF_PROF_UPPER}" STREQUAL "VTUNE")
+    TARGET_LINK_LIBRARIES(${libName} ittnotify)
+  ENDIF ()
+
+  TARGET_LINK_LIBRARIES(${libName} timing ${COMPOSE_LIBRARY} ${BLAS_LIBRARIES} ${LAPACK_LIBRARIES})
+
+  IF (HOMME_USE_KOKKOS)
+    TARGET_LINK_LIBRARIES(${libName} kokkos)
+  ENDIF ()
+
+  IF (NOT HOMME_USE_MKL)
+    IF (NOT HOMME_FIND_BLASLAPACK)
+      TARGET_LINK_LIBRARIES(${libName} lapack blas)
```

Comment:
It seems homme does not use blas/lapack anymore. Maybe we should try to remove this dependency (not in this PR).

Reply:
So it uses mkl, but not lapack? Interesting.

That said, if something is not used, I'm all for removing deps. Especially for libs like blas/lapack that tend to be available in system directories (with potential compiler version clash at link time).

Reply:
I'm not positive, but I think MKL and BLAS/LAPACK usage has always been 1-1 in Homme; there's no need for MKL if there's no need for BLAS/LAPACK. The fast DIRK solver is used in both C++ and F90 in all builds now, and I believe the original DIRK solver was the only user of MKL or BLAS/LAPACK before the fast solver was implemented for all builds. So that's why these dependencies aren't needed anymore. The code has ifdefs -- with no config-level settings -- that toggle the old solver, though. So it's conceivable someone might want to play around with that at some point. But in that case, it's reasonable to expect them to insert the dependencies manually using Extrae_LIBRARY.

Reply:
Grepping through the folder, I found two calls to DGEMM in `test_src/dcmip2016-supercell.F90`. I'm not sure if that's still used in practical tests, but it seems to be still compiled.

Reply:
Good find. DGESVD is also used in that file, so BLAS/LAPACK or the MKL interface to it are still needed. MKL isn't always available, so on at least some platforms we can expect to have to provide BLAS/LAPACK to get Homme to link.

Reply:
Just in case this comes up again and we look back at these comments: It just occurred to me that gllfvremap_mod also uses LAPACK, some QR factorization routines in particular.

Reply:
yes, but there is already a fix. Affected calls/functions/modules were wrapped in ifndef with HOMME_BUILD_WITHOUT_PIO, the change will be brought to master eventually.